src/rgw/CMakeLists.txt

```diff
@@ -434,49 +342,316 @@ target_include_directories(radosgw
+  driver/rados/rgw_rest_log.cc
+  driver/rados/rgw_rest_metadata.cc
+  driver/rados/rgw_rest_pubsub.cc
+  driver/rados/rgw_rest_realm.cc
+  driver/rados/rgw_rest_user.cc
+  driver/rados/rgw_role.cc
+  driver/rados/rgw_sal_rados.cc
+  driver/rados/rgw_service.cc
+  driver/rados/rgw_sync.cc
+  driver/rados/rgw_sync_counters.cc
+  driver/rados/rgw_sync_error_repo.cc
+  driver/rados/rgw_sync_module.cc
+  driver/rados/rgw_sync_module_aws.cc
+  driver/rados/rgw_sync_module_es.cc
+  driver/rados/rgw_sync_module_es_rest.cc
+  driver/rados/rgw_sync_module_log.cc
+  driver/rados/rgw_sync_trace.cc
+  driver/rados/rgw_tools.cc
+  driver/rados/rgw_trim_bilog.cc
+  driver/rados/rgw_trim_datalog.cc
+  driver/rados/rgw_trim_mdlog.cc
+  driver/rados/rgw_user.cc
+  driver/rados/rgw_zone.cc
+  driver/rados/config/impl.cc
+  driver/rados/config/period.cc
+  driver/rados/config/period_config.cc
+  driver/rados/config/realm.cc
+  driver/rados/config/store.cc
+  driver/rados/config/zone.cc
+  driver/rados/config/zonegroup.cc)
 
-# this is unsatisfying and hopefully temporary; ARROW should not be
-# part of radosgw_admin
-if(WITH_RADOSGW_ARROW_FLIGHT)
-  list(APPEND radosgw_admin_srcs rgw_flight.cc)
-endif(WITH_RADOSGW_ARROW_FLIGHT)
+add_library(rados_sal OBJECT ${rados_sal_srcs})
+target_include_directories(rados_sal
+  PUBLIC "services"
+  PUBLIC "${CMAKE_SOURCE_DIR}/src/rgw"
+  PUBLIC "${CMAKE_SOURCE_DIR}/src/rgw/driver/rados"
+  PUBLIC "${LUA_INCLUDE_DIR}"
+  PUBLIC "${CMAKE_SOURCE_DIR}/src/dmclock/support/src")
+
+if(WITH_RADOSGW_KAFKA_ENDPOINT)
+  # used by rgw_kafka.cc
+  target_link_libraries(rados_sal
+    PRIVATE
+      RDKafka::RDKafka)
+endif()
+if(WITH_RADOSGW_AMQP_ENDPOINT)
+  # used by rgw_amqp.cc
+  target_link_libraries(rados_sal
+    PRIVATE
+      RabbitMQ::RabbitMQ
+      OpenSSL::SSL)
+endif()
+if(WITH_OPENLDAP)
+  target_link_libraries(rados_sal
+    PRIVATE
+      OpenLDAP::OpenLDAP)
+endif()
+if(WITH_RADOSGW_LUA_PACKAGES)
+  target_link_libraries(rados_sal
+    PRIVATE Boost::filesystem StdFilesystem::filesystem)
+endif()
+
+if(WITH_LTTNG)
+  # rgw/rgw_op.cc includes "tracing/rgw_op.h"
+  # rgw/rgw_rados.cc includes "tracing/rgw_rados.h"
+  add_dependencies(rados_sal rgw_op-tp rgw_rados-tp)
+endif()
+
+if(WITH_JAEGER)
+  add_dependencies(rados_sal jaeger_base)
+  target_link_libraries(rados_sal PUBLIC jaeger_base)
+endif()
+
+target_link_libraries(rados_sal
+  PUBLIC 
+    rgw_common
+  PRIVATE
+    global
+    cls_2pc_queue_client
+    cls_cmpomap_client
+    cls_lock_client
+    cls_log_client
+    cls_otp_client
+    cls_refcount_client
+    cls_rgw_client
+    cls_rgw_gc_client
+    cls_timeindex_client
+    cls_user_client
+    cls_version_client
+    librados)
+set_target_properties(rados_sal PROPERTIES COMPILE_FLAGS "-fPIC")
+
+set(rgw_plugin_dir ${CEPH_INSTALL_PKGLIBDIR}/rgw_sal)
+
+set(rgw_sal_rados_srcs rgw_sal_rados_plugin.cc)
+add_library(rgw_sal_rados SHARED $<TARGET_OBJECTS:rados_sal> ${rgw_sal_rados_srcs})
```

Comment:
as long as the core rgw targets need to link against `rados_sal`, i think they should link directly against this shared library instead of dlopen'ing it

currently we're linking these `rados_sal` sources into both `radosgw` and `librgw_sal_rados.so` and loading both copies into the same process; that sounds problematic to me, especially if they're referring to global state

Reply:
I think we shouldn't be linking rados_sal into radosgw.

Reply:
This is the temporary/hack solution for the remaining undefined refs in Zone and ZoneGroup. When we get get Zone and ZoneGroup fixed then this goes away.   Not dlopening the plugin kinda goes against the concept of loadable plugins.

Reply:
Maybe there's another workaround, then?  I agree, loading these symbols twice is probably unsafe.

Reply:
> Not dlopening the plugin kinda goes against the concept of loadable plugins.

that's the final goal, but i'm looking for stepping-stones that let us merge this in pieces so you don't have to keep rebasing all of it. we could move *just* the dlopen part to a followup PR